lib/Dialect/FIRRTL/Transforms/CheckCombLoops.cpp

```diff
@@ -25,533 +26,454 @@
+          if (childVal.getDefiningOp() &&
+              (isa<RegOp, RegResetOp>(childVal.getDefiningOp()) ||
+               (isa<SubfieldOp, SubaccessOp, SubindexOp>(
+                    childVal.getDefiningOp()) &&
+                valRefersTo.find(childVal) == valRefersTo.end())))
+            continue;
+          if (!visited.contains(childVal))
+            dfsStack.push_back(childVal);
+          Node childNode;
+          bool isBackEdge = false;
+          // If the childVal is a sub, then check if it aliases with any of the
+          // predecessors (the visiting set).
+          forallRefersTo(childVal, [&](Node ref) {
+            if (visiting.contains(ref)) {
+              childNode = ref;
+              isBackEdge = true;
+              return failure();
+            }
+            return success();
+          });
 
-  /// Record all the leaf ground type values for an aggregate type value. The
-  /// Optional `fieldIdsToAdd` can be used to only add field ids that exist in
-  /// the set and ignore all other fields. This is used for MemOp ports, and it
-  /// will only contain 3 elements, so the copy shouldn't be a concern.
-  void gatherAggregateLeafs(Value val,
-                            SmallVector<std::pair<Value, Node>> &rootValues,
-                            llvm::SmallDenseSet<size_t> fieldIdsToAdd = {}) {
-    auto valType = val.getType().dyn_cast<FIRRTLBaseType>();
-    // Ignore RefType and GroundType Values.
-    if (!valType || valType.isGround())
-      return;
-    Node inputArg = Node();
-    if (val.isa<BlockArgument>())
-      inputArg = Node(val, 0);
-    SmallVector<Node> worklist;
-    worklist.push_back(Node(val, 0));
-    // Assign a unique storage location for the `val`, which is the aggregate
-    // base Value.
-    auto baseStorageLocId = maxStorageLocId;
-    valToStorageLocIdMap[val] = {baseStorageLocId};
-    storageLocIdToValuesMap[baseStorageLocId] = {val};
-    // Reserve unique storage location ids for all the subfields of `val`.
-    maxStorageLocId += valType.getMaxFieldID() + 1;
-    auto handleSub = [&](Value res, size_t fieldId) {
-      if (res.getType().cast<FIRRTLBaseType>().isGround()) {
-        if (!fieldIdsToAdd.empty())
-          if (!fieldIdsToAdd.contains(fieldId))
-            return;
-        // Compute the storage location id based on the fieldIndex.
-        auto storageLocId = baseStorageLocId + fieldId;
-        // The results of subaccess op can refer to multiple storage locations.
-        valToStorageLocIdMap[res].push_back(storageLocId);
-        storageLocIdToValuesMap[storageLocId].push_back(res);
-        if (!leafToBaseValMap.count(res)) {
-          leafToBaseValMap[res] = val;
-          if (inputArg.isValid())
-            inputArg.fieldId = fieldId;
-          rootValues.push_back({res, inputArg});
+          if (isBackEdge) {
+            // Comb Cycle Detected !!
+            lowLink[dfsVal] = std::min(lowLink[childVal], lowLink[dfsVal]);
```

Comment:
Is this safe re:simultaneous (unsequenced) accesses into the same DenseMap?  (it is if we can guarantee they're present at this point, it's not otherwise)

Reply:
Actually realized the entire discover time calculation is redundant, we report the error and exit after detecting the back edge, so removed the logic.